<a href="https://colab.research.google.com/github/ankitadhara28/Credit-Card-Fraud-Detection/blob/main/credit_Card_Fraud_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Credit Card Fraud Detection

## Introduction
Credit card fraud is a significant problem for financial institutions and consumers alike. This project aims to build a machine learning model to detect fraudulent credit card transactions using a given dataset.

## Dataset
The dataset used is `creditcard_2023.csv`, which contains anonymized transaction data. Key characteristics of the dataset include:
-   **Highly Imbalanced**: As observed, the dataset has a very skewed distribution between legitimate (Class 0) and fraudulent (Class 1) transactions. There are 18309 legitimate transactions and only 81 fraudulent transactions.
-   **Features**: The dataset contains 30 features (V1-V28, Amount, and id), with 'Class' being the target variable indicating whether a transaction is fraudulent (1) or legitimate (0).

## Methodology
The following steps were performed in the notebook:

1.  **Data Loading and Initial Exploration**:
    *   Loaded the `creditcard_2023.csv` into a pandas DataFrame.
    *   Displayed the first and last 5 rows to get a glimpse of the data.
    *   Used `credit_card_data.info()` to check data types and non-null counts.
    *   Checked for missing values using `credit_card_data.isnull().sum()`.
    *   Analyzed the distribution of the 'Class' variable using `value_counts()`.

2.  **Handling Imbalanced Data**:
    *   Separated legitimate and fraudulent transactions into two DataFrames (`legit` and `fraud`).
    *   Created a sample of legitimate transactions (`legit_sample`) with the same number of observations as fraudulent transactions (81 samples).
    *   Concatenated `legit_sample` and `fraud` to create a new, balanced dataset (`new_dataset`).

3.  **Data Splitting**:
    *   Separated features (`x`) from the target variable (`y`).
    *   Split the `new_dataset` into training and testing sets (`x_train`, `x_test`, `y_train`, `y_test`) using `train_test_split` with a `test_size` of 0.2, `stratify=y` to maintain the class distribution, and `random_state=2` for reproducibility.

4.  **Model Training**:
    *   Initialized a `LogisticRegression` model.
    *   Trained the model using the `x_train` and `y_train` datasets.

5.  **Model Evaluation**:
    *   Predicted on the training data and calculated the `accuracy_score`.
    *   Predicted on the test data and calculated the `accuracy_score`.

## Results

The Logistic Regression model achieved the following accuracy scores:
-   **Accuracy on training data**: 0.9767
-   **Accuracy on test data**: 0.9394

## Conclusion
The model shows good performance with high accuracy on both training and test datasets. The balancing of the dataset was crucial in enabling the model to learn from both classes effectively. Further improvements could involve exploring other classification algorithms, hyperparameter tuning, or more advanced feature engineering techniques.

In [7]:
# importing all the dependencies
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [8]:
# loading the dataset to the pandas DataFrame
credit_card_data = pd.read_csv("/content/creditcard_2023.csv")

In [9]:
# displaying first 5 rows of the dataset
credit_card_data.head()

,id,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0,-0.260648,-0.469648,2.496266,-0.083724,0.129681,0.732898,0.519014,-0.130006,0.727159,0.637735,-0.987020,0.293438,-0.941386,0.549020,1.804879,0.215598,0.512307,0.333644,0.124270,0.091202,-0.110552,0.217606,-0.134794,0.165959,0.126280,-0.434824,-0.081230,-0.151045,17982.10,0.0
1,1,0.985100,-0.356045,0.558056,-0.429654,0.277140,0.428605,0.406466,-0.133118,0.347452,0.529808,0.140107,1.564246,0.574074,0.627719,0.706121,0.789188,0.403810,0.201799,-0.340687,-0.233984,-0.194936,-0.605761,0.079469,-0.577395,0.190090,0.296503,-0.248052,-0.064512,6531.37,0.0
2,2,-0.260272,-0.949385,1.728538,-0.457986,0.074062,1.419481,0.743511,-0.095576,-0.261297,0.690708,-0.272985,0.659201,0.805173,0.616874,3.069025,-0.577514,0.886526,0.239442,-2.366079,0.361652,-0.005020,0.702906,0.945045,-1.154666,-0.605564,-0.312895,-0.300258,-0.244718,2513.54,0.0
3,3,-0.152152,-0.508959,1.746840,-1.090178,0.249486,1.143312,0.518269,-0.065130,-0.205698,0.575231,-0.752581,0.737483,0.592994,0.559535,-0.697664,-0.030669,0.242629,2.178616,-1.345060,-0.378223,-0.146927,-0.038212,-0.214048,-1.893131,1.003963,-0.515950,-0.165316,0.048424,5384.44,0.0
4,4,-0.206820,-0.165280,1.527053,-0.448293,0.106125,0.530549,0.658849,-0.212660,1.049921,0.968046,-1.203171,1.029577,1.439310,0.241454,0.153008,0.224538,0.366466,0.291782,0.445317,0.247237,-0.106984,0.729727,-0.161666,0.312561,-0.414116,1.071126,0.023712,0.419117,14278.97,0.0


In [10]:
# displaying last 5 rows
credit_card_data.tail()

,id,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
18386,18386,0.857988,-0.288022,0.495975,0.211239,0.503143,0.716816,0.554563,-0.118198,-0.160681,0.999267,0.124733,1.084228,-0.362396,1.265734,0.007188,0.436862,0.393389,-0.158408,-1.439516,-0.335855,-0.059347,0.182027,-0.106994,-0.475615,0.803368,0.255822,-0.258917,-0.103803,20269.23,0.0
18387,18387,0.950437,-0.951018,0.599227,-2.353135,-0.191018,0.213979,0.214060,-0.150530,1.042333,0.327260,0.096280,2.408187,0.721477,0.723804,0.905717,-0.533022,0.571632,1.573689,0.059785,-0.484363,-0.190511,-0.079367,-0.142690,0.124532,0.752516,-1.885803,-0.113027,-0.018291,19042.19,0.0
18388,18388,-0.379280,0.204033,0.405852,-0.356926,0.093057,0.492968,0.290180,0.178723,-0.052369,0.576023,-0.186651,1.910212,0.848084,1.581316,0.467337,0.439828,0.498113,0.500816,0.129773,-0.239093,0.017029,0.575236,-0.062849,-0.479217,-0.310835,-0.717239,-0.355992,0.069531,6975.19,0.0
18389,18389,0.001842,-0.374023,0.710190,-2.056370,1.678046,2.081933,0.471567,0.014155,0.395978,0.365183,-0.706091,0.421162,-0.012081,0.668933,1.000835,1.048479,0.120879,0.389597,-0.269786,-0.002891,-0.160530,-0.484981,-0.195650,2.324103,-0.287500,1.535573,-0.422965,-0.418794,22310.04,0.0
18390,18390,0.026208,-0.410460,2.016013,-0.896824,-0.005646,0.624816,0.404918,-0.065658,0.880602,0.330371,-0.495638,0.639787,-1.578406,0.608198,-1.075954,0.556608,0.400803,0.554120,0.636250,-0.183892,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [11]:
# dataset information
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18391 entries, 0 to 18390
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   id      18391 non-null  int64  
 1   V1      18391 non-null  float64
 2   V2      18391 non-null  float64
 3   V3      18391 non-null  float64
 4   V4      18391 non-null  float64
 5   V5      18391 non-null  float64
 6   V6      18391 non-null  float64
 7   V7      18391 non-null  float64
 8   V8      18391 non-null  float64
 9   V9      18391 non-null  float64
 10  V10     18391 non-null  float64
 11  V11     18391 non-null  float64
 12  V12     18391 non-null  float64
 13  V13     18391 non-null  float64
 14  V14     18391 non-null  float64
 15  V15     18391 non-null  float64
 16  V16     18391 non-null  float64
 17  V17     18391 non-null  float64
 18  V18     18391 non-null  float64
 19  V19     18391 non-null  float64
 20  V20     18391 non-null  float64
 21  V21     18390 non-null  float64
 22

In [12]:
# chceking no of misssing values in each column
credit_card_data.isnull().sum()

,0
id,0
V1,0
V2,0
V3,0
V4,0
V5,0
V6,0
V7,0
V8,0
V9,0


In [14]:
# distribution of legit transaction and fradulent transaction
credit_card_data['Class'].value_counts()

,count
Class,
0.0,18309
1.0,81


The dataset is highly unbalanced having unbalanced values for class 0 and class 1.

0 --> Normal Transaction

1 --> Fradulent Transaction

In [15]:
#seperating the data for analysis

legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]
print(legit.shape)
print(fraud.shape)

(18309, 31)
(81, 31)


In [16]:
# statistical measures of the data
legit.Amount.describe()

,Amount
count,18309.000000
mean,11904.577719
std,6923.745906
min,50.410000
25%,5905.730000
50%,11776.970000
75%,17909.570000
max,24039.880000


In [17]:
fraud.Amount.describe()

,Amount
count,81.000000
mean,11626.547284
std,6636.709826
min,925.610000
25%,5744.140000
50%,11281.610000
75%,17207.310000
max,23690.300000


In [18]:
# compare the values for both transaction
credit_card_data.groupby('Class').mean()

,id,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount
Class,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
0.0,9188.062264,0.321989,-0.403720,1.123283,-0.612274,0.277412,0.497389,0.438974,-0.144780,1.091200,0.583504,-0.333204,0.263518,0.838125,1.280370,-0.034873,0.574539,0.609318,0.381839,-0.305922,-0.141228,-0.131180,-0.154021,-0.051825,0.154115,0.127531,-0.008913,-0.203603,-0.084535,11904.577719
1.0,10649.666667,-0.951755,1.250584,-1.215923,1.223041,-0.959814,-1.240899,-1.010892,1.228171,-0.760299,-1.023191,1.323406,-1.226666,0.614255,-1.051873,-0.193487,-0.868173,-0.775553,-0.695578,0.049998,0.560552,0.082706,-0.378710,-0.321887,-0.403753,0.615130,0.351681,1.218448,0.108108,11626.547284


Build a Sample Dataset containing similiar distribution of normal transaction and Fradulent Transactions.

In [19]:
legit_sample = legit.sample(n=81)

In [21]:
# concatenating two data frame
new_dataset = pd.concat([legit_sample , fraud],axis = 0)

In [22]:
new_dataset.head()

,id,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
6998,6998,1.147544,-0.598617,0.573447,-0.772196,-0.033574,-0.164923,0.313755,-0.229974,0.898328,0.695237,-0.404880,-0.470714,1.172461,1.719304,-0.959344,0.837423,1.019327,-0.353017,0.289307,-0.149099,-0.100208,0.186152,-0.189929,0.794064,1.161356,-0.230296,-0.287760,-0.099095,3376.37,0.0
10160,10160,0.853807,-0.360462,1.702063,0.358661,0.038264,0.944279,0.179635,-0.065572,1.323328,0.714312,0.291572,-0.083895,1.276050,1.695807,-1.470006,0.779973,0.599120,0.380965,-1.302979,-0.430203,-0.163219,-0.083480,0.081943,0.314470,0.132350,-0.173496,-0.206936,-0.059220,5251.12,0.0
5246,5246,1.014343,-0.203245,0.399527,-0.049242,0.332621,-0.250100,0.577352,-0.240399,1.227249,0.297663,-0.122067,-0.253225,2.008348,1.174394,0.084224,0.618271,0.935286,0.619313,-0.800690,-0.277096,-0.196547,-0.374031,-0.140039,0.518135,1.120481,-0.967698,-0.249041,-0.008619,18884.68,0.0
5075,5075,0.923416,-0.517144,1.258002,-0.278855,0.020522,0.866995,0.154915,-0.075331,2.086267,0.416620,0.185272,-0.096597,1.107569,1.723218,-0.949144,0.569799,0.623556,0.504993,-0.375435,-0.357645,-0.180313,-0.172509,0.001625,-0.573452,0.162205,0.691167,-0.240285,-0.092580,19754.06,0.0
7617,7617,-1.100626,-7.295642,-0.372341,0.384057,-1.008659,1.706779,2.396670,-0.280995,1.049287,0.028102,0.346334,-0.221115,1.477686,2.225206,-1.059787,1.148004,1.286730,-0.102981,-1.113517,6.855674,0.768155,-1.575396,-3.081670,-0.197489,-2.085366,-1.688562,-0.948277,1.568333,7606.44,0.0


In [23]:
new_dataset.tail()

,id,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
17317,17317,-2.521749,3.153763,-2.417978,1.262006,-3.108947,-2.798076,-2.368268,5.389400,-1.085033,-1.199577,0.963508,-0.916017,1.230939,-0.716694,-0.165140,-1.109864,-1.335300,-1.418689,0.390728,1.471750,0.570627,-1.751248,-1.228181,0.360898,3.446097,-0.488977,1.624072,0.835026,925.61,1.0
17366,17366,-2.551425,3.207646,-2.443510,1.268499,-3.157705,-2.825894,-2.396545,5.530965,-1.076135,-1.188761,0.932615,-0.892054,1.308520,-0.677456,-0.173626,-1.088250,-1.319274,-1.414041,0.357399,1.480063,0.571913,-1.794574,-1.271039,0.369819,3.543782,-0.487599,1.586523,0.816186,8102.86,1.0
17407,17407,-2.580623,3.261359,-2.468736,1.274981,-3.206065,-2.853719,-2.424652,5.672931,-1.067219,-1.177907,0.901360,-0.867605,1.386058,-0.637179,-0.182094,-1.066354,-1.303155,-1.409387,0.323955,1.488318,0.573191,-1.837772,-1.313817,0.378760,3.641667,-0.486217,1.549170,0.797321,19918.66,1.0
17453,17453,-2.609364,3.314906,-2.493666,1.281451,-3.254039,-2.881551,-2.452595,5.815292,-1.058284,-1.167012,0.869728,-0.842647,1.463552,-0.595792,-0.190545,-1.044166,-1.286941,-1.404726,0.290393,1.496519,0.574464,-1.880846,-1.356513,0.387721,3.739753,-0.484832,1.512010,0.778434,1111.14,1.0
17480,17480,-2.637662,3.368287,-2.518308,1.287910,-3.301638,-2.909392,-2.480375,5.958040,-1.049329,-1.156078,0.837708,-0.817155,1.541006,-0.553212,-0.198978,-1.021677,-1.270630,-1.400058,0.256709,1.504670,0.575730,-1.923798,-1.399126,0.396701,3.838035,-0.483444,1.475041,0.759524,19813.32,1.0


In [24]:
new_dataset['Class'].value_counts()

,count
Class,
0.0,81
1.0,81


In [26]:
new_dataset.groupby('Class').mean()

,id,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount
Class,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
0.0,8867.555556,0.289150,-0.464373,1.152440,-0.586421,0.171114,0.499060,0.372968,-0.166393,1.283463,0.602742,-0.416962,0.158296,0.686474,1.265920,-0.272447,0.478454,0.711872,0.331278,-0.269112,-0.131106,-0.078072,-0.243703,-0.025576,0.191697,0.115641,-0.138731,-0.247416,0.020957,12326.765926
1.0,10649.666667,-0.951755,1.250584,-1.215923,1.223041,-0.959814,-1.240899,-1.010892,1.228171,-0.760299,-1.023191,1.323406,-1.226666,0.614255,-1.051873,-0.193487,-0.868173,-0.775553,-0.695578,0.049998,0.560552,0.082706,-0.378710,-0.321887,-0.403753,0.615130,0.351681,1.218448,0.108108,11626.547284


Splitting data into features and target

In [27]:
x = new_dataset.drop(columns = 'Class',axis = 1)
y = new_dataset['Class']
print(x)
print(y)

          id        V1        V2  ...       V27       V28    Amount
6998    6998  1.147544 -0.598617  ... -0.287760 -0.099095   3376.37
10160  10160  0.853807 -0.360462  ... -0.206936 -0.059220   5251.12
5246    5246  1.014343 -0.203245  ... -0.249041 -0.008619  18884.68
5075    5075  0.923416 -0.517144  ... -0.240285 -0.092580  19754.06
7617    7617 -1.100626 -7.295642  ... -0.948277  1.568333   7606.44
...      ...       ...       ...  ...       ...       ...       ...
17317  17317 -2.521749  3.153763  ...  1.624072  0.835026    925.61
17366  17366 -2.551425  3.207646  ...  1.586523  0.816186   8102.86
17407  17407 -2.580623  3.261359  ...  1.549170  0.797321  19918.66
17453  17453 -2.609364  3.314906  ...  1.512010  0.778434   1111.14
17480  17480 -2.637662  3.368287  ...  1.475041  0.759524  19813.32

[162 rows x 30 columns]
6998     0.0
10160    0.0
5246     0.0
5075     0.0
7617     0.0
        ... 
17317    1.0
17366    1.0
17407    1.0
17453    1.0
17480    1.0
Name: Class, Len

Split data into train and test data

In [28]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2, stratify = y, random_state=2)

In [30]:
print(x.shape , x_train.shape , x_test.shape)

(162, 30) (129, 30) (33, 30)


Using Logistic Regression model for binary data

In [32]:
model = LogisticRegression()

In [33]:
# training the logistic regression model with training data
model.fit(x_train,y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

Model Evaluation based on accuracy score

In [35]:
# accuracy score on training data
x_train_prediction = model.predict(x_train)
training_data_accuracy = accuracy_score(x_train_prediction,y_train)
print('Accuracy on training data : ',training_data_accuracy)

Accuracy on training data :  0.9767441860465116


In [36]:
# accuracy on test data
x_test_prediction = model.predict(x_test)
test_data_accuracy = accuracy_score(x_test_prediction, y_test)
print('Accuracy on test data:', test_data_accuracy)

Accuracy on test data: 0.9393939393939394


# Credit Card Fraud Detection

![Python Version](https://img.shields.io/badge/Python-3.x-blue.svg)
![Scikit-learn Version](https://img.shields.io/badge/Scikit--learn-0.24%2B-orange.svg)

## Table of Contents
- [Introduction](#introduction)
- [Dataset](#dataset)
- [Methodology](#methodology)
- [Results](#results)
- [Future Improvements](#future-improvements)
- [Installation](#installation)
- [Usage](#usage)
- [Contributing](#contributing)
- [License](#license)

## Introduction
This project focuses on building a machine learning model to detect fraudulent credit card transactions. Credit card fraud is a significant challenge for financial institutions and consumers, leading to substantial financial losses and erosion of trust. By leveraging machine learning, we aim to accurately identify suspicious transactions, thereby minimizing fraudulent activities.

## Dataset
The dataset used for this project is `creditcard_2023.csv`, obtained from a publicly available source (e.g., Kaggle). It contains anonymized transaction data with the following key characteristics:

-   **Highly Imbalanced**: The dataset exhibits a severe class imbalance, with a significantly higher number of legitimate transactions (Class 0) compared to fraudulent ones (Class 1). Specifically, there are 18309 legitimate transactions and only 81 fraudulent transactions.
-   **Features**: The dataset comprises 30 features, primarily anonymized principal components (`V1` through `V28`), `Amount`, and `id`. The target variable, `Class`, indicates whether a transaction is legitimate (0) or fraudulent (1).

## Methodology
The fraud detection model was developed following these key steps:

1.  **Data Loading and Initial Exploration**:
    *   Loaded the `creditcard_2023.csv` file into a pandas DataFrame.
    *   Performed initial checks on the data, including viewing the first/last rows, inspecting data types (`.info()`), and identifying missing values (`.isnull().sum()`).
    *   Analyzed the distribution of the `Class` variable to understand the extent of class imbalance.

2.  **Handling Imbalanced Data**:
    *   To address the severe class imbalance, a balanced dataset was created. Legitimate transactions (Class 0) and fraudulent transactions (Class 1) were separated.
    *   A sample of legitimate transactions (`legit_sample`) was randomly selected to match the number of fraudulent transactions (81 samples).
    *   These two subsets (`legit_sample` and `fraud`) were then concatenated to form a new, balanced dataset (`new_dataset`).

3.  **Data Splitting**:
    *   The balanced dataset was split into features (`x`, dropping the 'Class' column) and the target variable (`y`, the 'Class' column).
    *   The data was further divided into training and testing sets (`x_train`, `x_test`, `y_train`, `y_test`) using `train_test_split`. A `test_size` of 0.2 (20% for testing) was used, with `stratify=y` to ensure that the class distribution is preserved in both training and testing sets, and `random_state=2` for reproducibility.

4.  **Model Training**:
    *   A Logistic Regression model (`LogisticRegression()`) was initialized.
    *   The model was trained using the `x_train` and `y_train` datasets.

5.  **Model Evaluation**:
    *   The trained model's performance was evaluated using `accuracy_score` on both the training and testing datasets.

## Results

The Logistic Regression model achieved the following accuracy scores:

-   **Accuracy on training data**: `0.9767` (approximately 97.67%)
-   **Accuracy on test data**: `0.9394` (approximately 93.94%)

These results indicate that the model performs well on both seen and unseen data, demonstrating its ability to generalize. The balancing technique applied was effective in allowing the model to learn from the minority class.

## Future Improvements

To further enhance the model's performance and robustness, consider the following:

-   **Exploring other algorithms**: Evaluate other classification algorithms such as Random Forest, Gradient Boosting, or Support Vector Machines.
-   **Hyperparameter Tuning**: Optimize the hyperparameters of the chosen model using techniques like GridSearchCV or RandomizedSearchCV.
-   **Advanced Resampling Techniques**: Investigate more sophisticated resampling methods (e.g., SMOTE, ADASYN) for handling imbalanced datasets.
-   **Feature Engineering**: Create new features from existing ones that might provide more predictive power.
-   **Cross-Validation**: Implement k-fold cross-validation for a more robust evaluation of the model's performance.

## Installation

To run this notebook, you will need the following Python libraries. You can install them using pip:

```bash
pip install pandas numpy scikit-learn
```

## Usage

1.  Clone this repository:
    ```bash
    git clone https://github.com/yourusername/credit-card-fraud-detection.git
    cd credit-card-fraud-detection
    ```
2.  Place the `creditcard_2023.csv` dataset in the same directory as the notebook.
3.  Open the `Credit Card Fraud Detection.ipynb` notebook in Google Colab or Jupyter Notebook.
4.  Run all cells to replicate the analysis and model training.

## Contributing

Contributions are welcome! If you have suggestions for improvements or new features, please open an issue or submit a pull request.

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.